uvicorn/supervisors/manager.py

```diff
@@ -0,0 +1,157 @@
+            "int",  # Interrupt signal generated by Ctrl+C
+            "term",  # Termination signal
+            "winch",  # Window size change signal
+            "chld",  # Child process terminated, stopped, or continued
+        )
+    }
+
+    # TODO(Marcelo): This should be converted into a CLI option.
+    GRACEFUL_TIMEOUT = 30
+
+    def __init__(self, config: Config, target: Target, sockets: List[socket]) -> None:
+        self.config = config
+        self.target = target
+        self.sockets = sockets
+        self.processes: List[SpawnProcess] = []
+        self.sig_queue = mp.Queue()
+
+    def run(self) -> None:
+        self.start()
+
+        try:
+            self.spawn_processes()
+
+            while True:
+                try:
+                    sig = self.sig_queue.get(timeout=0.25)
+                except queue.Empty:
+                    self.reap_processes()
+                    self.spawn_processes()
+                    continue
+
+                if sig not in self.SIGNALS.keys():
+                    logger.info("Ignoring unknown signal: %d", sig)
+                    continue
+
+                handler = self.signal_handler(sig)
+                if handler is None:
+                    logger.info("Unhandled signal: %s", self.SIGNALS.get(sig))
+                    continue
+
+                handler()
+        except StopIteration:
+            self.halt()
+        except Exception:
+            logger.info("Unhandled exception in main loop", exc_info=True)
+            self.stop(signal.SIGTERM)
+            self.halt(ExitCode.DEFAULT_FAILURE)
+
+    def start(self) -> None:
+        self.pid = os.getpid()
+        logger.info("Started manager process [%d]", self.pid)
+        self.init_signals()
+
+    def spawn_processes(self) -> None:
+        for _ in range(self.config.workers - len(self.processes)):
+            self.spawn_process()
+
+    def spawn_process(self) -> None:
+        process = get_subprocess(self.config, target=self.target, sockets=self.sockets)
+        process.start()
+        self.processes.append(process)
+
+    def init_signals(self) -> None:
+        for s in self.SIGNALS.keys():
+            signal.signal(s, self._signal)
+        signal.signal(signal.SIGCHLD, self.handle_chld)
+
+    def _signal(self, sig: signal.Signals, frame: FrameType) -> None:
+        self.sig_queue.put(sig)
+
+    def handle_int(self) -> None:
+        self.stop(signal.SIGINT)
+        raise StopIteration
+
+    def handle_term(self) -> None:
+        self.stop(signal.SIGTERM)
+        raise StopIteration
+
+    def handle_chld(self, sig: signal.Signals, frame: FrameType) -> None:
+        self.reap_processes()
+
+    def signal_handler(self, sig: signal.Signals) -> Optional[Callable[..., None]]:
+        sig_name = self.SIGNALS.get(sig)
+        return getattr(self, f"handle_{sig_name}", None)
+
+    def reap_processes(self) -> None:
+        for process in self.processes:
+            if not process.is_alive():
```

Comment:
I need to further check if the `wait()` here is enough to reap! It worked on the cases I've tested, but I'm not sure if there are edge cases!